### PR TITLE
feat: add adaptive cycle termination

### LINKED ISSUE #14: [Feature]: Adaptive cycle termination — automatic convergence detection
## Problem Statement

Users must manually specify `num_cycles` with no feedback on whether additional cycles are actually improving robustness. Once robustness saturates (which typically happens around cycle 3-5 for most models), every subsequent cycle wastes compute without meaningful gains. There's currently no mechanism for the pipeline to introspect its own progress and terminate early.

This matters because: (a) GPU time is expensive, especially on cloud instances, (b) users can't know the optimal cycle count without running full sweeps first, and (c) the 4-phase loop is the core differentiator - it should be smart enough to self-terminate.

**Relevant files:** `nightmarenet/pipeline.py` (orchestrator), `nightmarenet/training/trainer.py` (training loop), `nightmarenet/evaluation/metrics.py` (robustness scoring), `configs/default.yaml` (config surface).

## Proposed Solution

Automatically detect convergence and terminate training early when robustness plateaus. The approach:

- Introduce a lightweight per-cycle robustness probe (separate from the full `Evaluator.evaluate()`) that runs on a small fixed subset (~50 samples) at a single distortion strength after each cycle's Compress phase completes
- Track the delta in robustness score between consecutive cycles
- Auto-terminate when delta falls below a configurable threshold for N consecutive cycles
- Expose via config while preserving `num_cycles` as a hard upper-bound cap
- Report convergence metadata ("Converged after N cycles, delta = X%") in the final evaluation output

**Files affected:**
- `nightmarenet/training/trainer.py` - cycle-end event emission + stop mechanism
- `nightmarenet/pipeline.py` - convergence detection logic in the progress callback
- `nightmarenet/evaluation/metrics.py` - new `quick_robustness_score()` helper
- `configs/default.yaml` - new config keys

**Design decisions:**
- Per-cycle probe is intentionally lightweight (single strength, small sample). Full `Evaluator.evaluate()` is too expensive to run every cycle.
- Convergence logic lives in Pipeline (not Trainer), keeping Trainer focused on phase execution
- `num_cycles` becomes a hard cap rather than exact count when auto-terminate is enabled

**New dependencies:** None.

## Alternatives Considered

- Running full `Evaluator.evaluate()` per cycle - rejected, too expensive (full distortion sweep at 9 strengths)
- Loss-based early stopping only (already exists via `early_stopping` config) - doesn't capture robustness; a model can have low loss but poor adversarial robustness
- User-defined callbacks for custom termination - too much burden on users; convergence detection should be built-in

## Acceptance Criteria

- [ ] `quick_robustness_score()` evaluates model on a fixed subset at a single strength and returns a scalar score
- [ ] Trainer emits a `cycle_end` event after the Compress phase of each cycle
- [ ] Pipeline intercepts `cycle_end`, runs the quick probe, tracks deltas
- [ ] Training stops early when delta < threshold for `convergence_patience` consecutive cycles
- [ ] Final report includes convergence info (cycles completed, final delta)
- [ ] Config keys `auto_terminate`, `convergence_threshold`, `convergence_patience` added to `configs/default.yaml`
- [ ] Existing behavior unchanged when `auto_terminate: false` (backward compatible)
- [ ] Tests added covering: convergence detection triggers correctly, patience respected, hard cap still honored
- [ ] Documentation updated in default.yaml comments

## Area

Core training loop, Evaluation/metrics

## Priority

Important

### PR DESCRIPTION
**Pull request by malabikaparia**
## Summary

Implement adaptive cycle termination based on convergence of a lightweight robustness probe evaluated after each training cycle.

## Motivation

This adds adaptive training termination by monitoring robustness convergence, reducing unnecessary training cycles while preserving existing behavior when disabled.

Closes #14 

## Changes

- Added `quick_robustness_score()` for lightweight robustness evaluation on a fixed dataset subset.
- Added `cycle_end` event emission and graceful external stop support in `Trainer`.
- Added adaptive convergence detection in `Pipeline`.
- Added configurable training settings:
  - `auto_terminate`
  - `convergence_threshold`
  - `convergence_patience`
- Added convergence metadata to the evaluation report.
- Extracted adaptive convergence handling into a private helper method to isolate the convergence logic for unit testing without changing the training flow.
- Added unit tests covering adaptive termination behavior.

## Acceptance Criteria

- [x] `quick_robustness_score()` evaluates model on a fixed subset at a single strength and returns a scalar score.
- [x] `Trainer` emits a `cycle_end` event after the Compress phase of each cycle.
- [x] `Pipeline` intercepts `cycle_end`, runs the quick probe, tracks deltas.
- [x] Training stops early when delta < threshold for `convergence_patience` consecutive cycles.
- [x] Final report includes convergence info(cycles completed, final delta)
- [x] Config keys `auto_terminate`, `convergence_threshold`, and `convergence_patience` added to `configs/default.yaml`.
- [x] Existing behavior remains unchanged when `auto_terminate: false`. (backward compatible)
- [x] Tests added covering: convergence detection triggers correctly, patience respected, hard cap still honored
- [x] Documentation updated in default.yaml comments.

## Type

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] Feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would break existing behavior)
- [x] Refactor (no functional change)
- [ ] Documentation
- [x] Tests

## Pre-submission Checklist

- [x] I have **starred** this repository
- [x] I have **followed** [@Adit-Jain-srm](https://github.com/Adit-Jain-srm)
- [x] I have read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md)

## Quality Checklist

- [x] `ruff check nightmarenet/ tests/` passes with 0 errors
- [x] `mypy nightmarenet/ --ignore-missing-imports` passes
- [x] `pytest tests/` — all tests pass (445+)
- [x] Added tests for new functionality (if applicable)
- [x] Updated documentation (if applicable)
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)
- [x] All acceptance criteria from the linked issue are satisfied (or exceptions noted above)

## Screenshots (if UI change)

N/A

### CODE REVIEW

**Comment by Adit-Jain-srm** [comment]
## Review: PR #55 - Adaptive Cycle Termination

### Status

- **Merge conflicts** - PR is in CONFLICTING state. CI (Tests & Lint) hasn't run. Resolve conflicts with main before review can complete.
- Architecture follows the guidance from the [issue #14 thread](https://github.com/Adit-Jain-srm/NightmareNet/issues/14#issuecomment-4879178040) closely.

---

### Issue #14 Acceptance Criteria

| Criterion | Status |
|---|---|
| `quick_robustness_score()` returns scalar on fixed subset | Done |
| Trainer emits `cycle_end` after Compress phase | Done |
| Pipeline intercepts `cycle_end`, runs probe, tracks deltas | Done |
| Stops when delta < threshold for `convergence_patience` cycles | Done |
| Final report includes convergence info | Done |
| Config keys in `configs/default.yaml` | Done |
| Backward compatible when `auto_terminate: false` | Done (tested) |
| Tests: convergence triggers, patience respected, hard cap | Done (3 tests) |
| Documentation in default.yaml comments | Done |

All 9 criteria addressed.

---

### Required Actions

#### 1. Resolve merge conflicts

PR is `CONFLICTING`. Main has moved since this branch was created (PR #43 merged webhook changes to pipeline.py and trainer.py). Merge main and resolve.

#### 2. Circular dependency: `metrics.py` imports from `trainer.py`

```python
# nightmarenet/evaluation/metrics.py (new import):
from nightmarenet.training.trainer import _tokenize_dataset
```

This creates an evaluation-depends-on-training dependency (wrong direction). `_tokenize_dataset` is a utility function that doesn't belong exclusively to trainer. Move it to `nightmarenet/utils/` or `nightmarenet/data/` and import from there in both metrics.py and trainer.py. Alternatively, inline the tokenization logic in `quick_robustness_score` (it's only ~10 lines).

#### 3. `_stop_requested` check happens at phase boundary, not cycle boundary

In trainer.py:
```python
for cycle, phase, num_epochs in self.scheduler:
    if self._stop_requested:
        break
```

This breaks at the START of the next phase, not the next cycle. If `request_stop()` is called during the compress phase's `on_progress` callback, the loop breaks before the next cycle's wake starts - which is correct. But if for some reason it's called mid-cycle (e.g., during dream), it would break before nightmare, leaving an incomplete cycle. Since `_handle_cycle_end` only fires after compress, this is fine in practice. Just noting the asymmetry.

#### 4. `apply_text_distortions` as the probe distortion function

```python
from nightmarenet.distortions.text import apply_text_distortions
```

The pipeline uses this as the fixed distortion function for the convergence probe. Verify this matches the signature `(text: str, strength: float) -> str` that `quick_robustness_score` expects. If `apply_text_distortions` has a different signature (e.g., requires `seed` or returns something else), the probe will crash at runtime. CI can't catch this without an integration test that actually runs the probe.

#### 5. Missing blank line after class attribute block

```python
        self._cycles_completed = 0
    # ------------------------------------------------------------------
    # Private helpers
```

Ruff might flag this (no blank line between attribute assignments and section comment). Minor, but will fail CI if ruff is strict about it.

---

### What's Solid

- Clean separation: convergence logic in Pipeline (`_handle_cycle_end`), stop mechanism in Trainer (`request_stop`), probe in metrics (`quick_robustness_score`)
- `cycle_end` event fires only after compress phase (correct cycle boundary)
- Tests mock the probe and verify patience + threshold logic independently
- Config is backward compatible (defaults to `false`)
- Convergence metadata included in evaluation comparison output

---

### Verdict

Resolve conflicts (#1) and fix the circular import (#2). Once CI passes, this is ready to merge.

**Review by Adit-Jain-srm — CHANGES_REQUESTED** [review]
Implement these and it's ready to merge @malabikaparia

**Review by Adit-Jain-srm — APPROVED** [review]
LGTM!
